### PR TITLE
matchers: implement on_no_match for custom matchers

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>
Commit Message: matchers: implement on_no_match for custom matchers
Risk Level: low (new feature)
Testing: unit
Docs Changes: none
Release Notes:

### CODE REVIEW

**Review by rojkov — COMMENTED** [review]
Thank you! Added one small nit.

File: source/common/matcher/matcher.h [hunk 1]
```diff
@@ -211,8 +213,6 @@ class MatchTreeFactory : public OnMatchFactory<DataType> {
             std::make_pair(children.first, *MatchTreeFactory::createOnMatch(children.second)));
       }
 
-      auto on_no_match = createOnMatch(matcher.on_no_match());
-
```
Comment (rojkov) [inline source/common/matcher/matcher.h:215]: Nit: the style guide encourages declaring local variables as close to their first use as possible.

Comment (kyessenov) [reply source/common/matcher/matcher.h:215]: It's used in both branches of `switch`. We'd have to duplicate the line to make it any closer. Is that recommended?

Comment (rojkov) [reply source/common/matcher/matcher.h:215]: Never mind then. I missed it's used in greater scope now.

**Review by rojkov — APPROVED** [review]
Looks good to me. Thanks!

**Review by snowp — APPROVED** [review]
Thanks!

**Pull request merged**